src/Groups/matrices/forms.jl

```diff
@@ -467,3 +468,399 @@ function issingular(f::SesquilinearForm{T}) where T
    return GAP.Globals.IsSingularForm(f.X)
 end
+
+
+
+
+########################################################################
+#
+# From group to form #TODO: there are different approaches. Which is the best?
+#
+########################################################################
+
+# TODO 1st approach: brute force calculation
+# Algorithm furnished by Thomas Breuer, Aachen University
+# extended to quadratic by Giovanni De Franceschi, TU Kaiserslautern
+# WARNING: huge linear system !!
+
+"""
+    invariant_bilinear_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of bilinear forms preserved by the group `G`.
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+"""
+function invariant_bilinear_forms(G::MatrixGroup{S,T}) where {S,T}
+   F = base_ring(G)
+   n = degree(G)
+   M = T[]
+   for mat in gens(G)
+      mmat = mat^-1
+      MM = zero_matrix(F,n^2,n^2)
+      for i in 1:n, j in 1:n, k in 1:n
+         MM[n*(i-1)+j,n*(k-1)+j] += mat[i,k]
+         MM[n*(i-1)+j,n*(i-1)+k] -= mmat[j,k]
+      end
+      push!(M,MM)
+   end
+
+   r,K = nullspace(block_matrix(length(M),1,M))
+   
+   return [matrix(F,n,n,[K[i,j] for i in 1:n^2]) for j in 1:r]
+end
+
+"""
+    invariant_sesquilinear_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of sesquilinear non-bilinear forms preserved by the group `G`. It works only if `base_ring(G)` has even degree.
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+"""
+function invariant_sesquilinear_forms(G::MatrixGroup{S,T}) where {S,T}
+   F = base_ring(G)
+   @assert iseven(degree(F)) "Base ring has no even degree"
+   n = degree(G)
+   M = T[]
+   for mat in gens(G)
+      mmat = map(y->frobenius(y,div(degree(F),2)),(mat.elm)^-1)
+      MM = zero_matrix(F,n^2,n^2)
+      for i in 1:n, j in 1:n, k in 1:n
+         MM[n*(i-1)+j,n*(k-1)+j] += mat[i,k]
+         MM[n*(i-1)+j,n*(i-1)+k] -= mmat[j,k]
+      end
+      push!(M,MM)
+   end
+
+   r,K = nullspace(block_matrix(length(M),1,M))
+   
+   return [matrix(F,n,n,[K[i,j] for i in 1:n^2]) for j in 1:r]
+end
+
+"""
+    invariant_quadratic_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of quadratic forms preserved by the group `G`.
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+"""
+function invariant_quadratic_forms(G::MatrixGroup{S,T}) where {S,T}
+   F = base_ring(G)
+   n = degree(G)
+   M = T[]
+   for mat in gens(G)
+      MM = zero_matrix(F,div(n*(n+1),2),div(n*(n+1),2))
+      for i in 1:n
+      for p in 1:n, q in p:n
+         MM[div((2*n-i+2)*(i-1),2)+1, div((2*n-p+2)*(p-1),2)+q-p+1] = mat[i,p]*mat[i,q]
+      end
+      for j in i+1:n, p in 1:n, q in p:n
+         MM[div((2*n-i+2)*(i-1),2)+j-i+1, div((2*n-p+2)*(p-1),2)+q-p+1] = mat[i,p]*mat[j,q]+mat[j,p]*mat[i,q]
+      end
+      end
```

Comment:
So perhaps reformat and merge these, and give names to some of the intermediate value? Something like this (please double check I didn't mess it up)
```suggestion
      for i in 1:n
         row = div((2*n-i+2)*(i-1),2)+1
         for p in 1:n, q in p:n
            col = div((2*n-p+2)*(p-1),2)+q-p+1
            MM[row, col] = mat[i,p]*mat[i,q]
            for j in i+1:n
               MM[row+j-i, col] = mat[i,p]*mat[j,q]+mat[j,p]*mat[i,q]
            end
         end
      end
```

Reply:
Done. It works.